src/Features/CSharp/Portable/BraceCompletion/InterpolatedStringBraceCompletionService.cs

```diff
@@ -5,87 +5,83 @@
+    protected override bool IsValidOpenBraceTokenAtPosition(SourceText text, SyntaxToken token, int position)
+        => IsValidOpeningBraceToken(token) && token.Span.End - 1 == position;
 
-        protected override bool IsValidOpeningBraceToken(SyntaxToken leftToken)
-            => leftToken.Kind() is SyntaxKind.InterpolatedStringStartToken or SyntaxKind.InterpolatedVerbatimStringStartToken;
-
-        protected override bool IsValidClosingBraceToken(SyntaxToken rightToken)
-            => rightToken.IsKind(SyntaxKind.InterpolatedStringEndToken);
+    /// <summary>
+    /// Returns true when the input position could be starting an interpolated string if opening quotes were typed.
+    /// </summary>
+    public static bool IsPositionInInterpolatedStringContext(ParsedDocument document, int position, CancellationToken cancellationToken)
+    {
+        var text = document.Text;
 
-        protected override bool IsValidOpenBraceTokenAtPosition(SourceText text, SyntaxToken token, int position)
-            => IsValidOpeningBraceToken(token) && token.Span.End - 1 == position;
+        var start = position - 1;
+        if (start < 0)
+            return false;
 
-        /// <summary>
-        /// Returns true when the input position could be starting an interpolated string if opening quotes were typed.
-        /// </summary>
-        public static bool IsPositionInInterpolatedStringContext(ParsedDocument document, int position, CancellationToken cancellationToken)
+        // Check if the user is typing an interpolated or interpolated verbatim string.
+        // If the preceding character(s) are not '$' or '$@' then we can't be starting an interpolated string.
+        if (text[start] == '@')
         {
-            var text = document.Text;
-
-            var start = position - 1;
-            if (start < 0)
+            // must have $@.  otherwise this is some other @ construct.
+            if (start == 0 || text[start - 1] != '$')
                 return false;
 
-            // Check if the user is typing an interpolated or interpolated verbatim string.
-            // If the preceding character(s) are not '$' or '$@' then we can't be starting an interpolated string.
-            if (text[start] == '@')
-            {
-                if (--start < 0)
-                    return false;
-            }
-
-            if (text[start] != '$')
-                return false;
-
-            // Verify that we are actually in an location allowed for an interpolated string.
-            var token = document.Root.FindToken(start);
-            if (token.Kind() is not SyntaxKind.InterpolatedStringStartToken and
-                                not SyntaxKind.InterpolatedVerbatimStringStartToken and
-                                not SyntaxKind.StringLiteralToken and
-                                not SyntaxKind.IdentifierToken)
```

Comment:
prior logic was too much syntax tree based.  this made it fall over when the tree wasn't well formed (like just having a naked `$` in an expression).

The fix is to smartly moved before those characters.  We then know we must have a string  (since we have a `$` and we know the user is typing `"`).   We then use the existing helpers to then ask "is this a legal place for such a string".  

Importantly, we're only looking at the tree *prior* to the current token we're writing.  we're not trying to examine the in-progress token itself directly.